Fixed(api): changed administrator invitation service

Hrushon opened the pull request:
Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.

[bug1](https://www.notion.so/128-d3f6c10dabe84edf908d97704d38645e)
[bug2](https://www.notion.so/130-invitation_id-fa2722af498043e491dee66b31e29d8f)

Hrushon committed:
fix(api): changed administrator invitation service

Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.

AntonRev commented on src/core/services/authentication_service.py:
```diff
@@ -91,3 +92,9 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
             administrator=administrator,
         )
+
+    async def check_administrator_not_expert(self, token: str) -> None:
```
Название функции не совсем корректное. Если в будущем добавятся роли, по названию функции будет непонятно что проверяется

AntonRev commented on src/core/services/authentication_service.py:
```diff
@@ -91,3 +92,9 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
+        """Проверяет, что администратор не эксперт."""
+        administrator = await self.get_current_active_administrator(token.credentials)
+        if administrator is not Administrator.Role.ADMINISTRATOR:
+            raise AdministratorCheckError()
```
`AdministratorCheckError` из названия тоже не ясно из-за какой проверки вызывается исключение.

Hrushon committed:
fix(api): changed administrator invitation service

Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.

Ivan-Skvortsov commented on src/api/routers/administrator_invitation.py:
```diff
@@ -112,7 +112,7 @@ async def deactivate_invitation(
         self, invitation_id: UUID, token: HTTPAuthorizationCredentials = Depends(HTTPBearer())
     ) -> Any:
-        await self.authentication_service.get_current_active_administrator(token.credentials)
+        await self.authentication_service.check_administrator_have_rights(token.credentials)
```
Может быть, есть смысл просто добавить в `get_current_active_administrator` какой-нибудь флаг, по которому проверяется эксперт это или админ и сделать его по дефолту False? Например `is_admin`?

Hrushon replied:
точно, не додумался

AntonZelinsky replied:
Только просто назвать `get_current_administrator`, то что нельзя получить не активного админа можно опустить.

Ivan-Skvortsov commented on src/core/exceptions.py:
```diff
@@ -299,3 +300,9 @@ class AdministratorSelfBlockError(ApplicationException):
         self.detail = "Вы не можете заблокировать себя."
+
+
+class AdministratorCheckRightsError(ApplicationException):
```
Предлагаю назвать это как-нибудь вроде AdministratorInviteError с описанием "У Вас нет прав на отправку приглашений другим администраторам"

Ivan-Skvortsov commented on src/core/services/authentication_service.py:
```diff
@@ -91,3 +92,9 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
             administrator=administrator,
         )
+
+    async def check_administrator_have_rights(self, token: str) -> None:
```
Надо бы проверки на права админа также добавить во все эндпойнты, связанные с приглашениями

Hrushon committed:
fix(api): changed administrator invitation service

Version 3.
Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.

Hrushon committed:
fix(api): changed administrator invitation service

Version 3.
Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.

AntonRev commented on src/core/services/authentication_service.py:
```diff
@@ -82,6 +82,8 @@ async def get_current_active_administrator(self, token: str) -> Administrator:
         administrator = await self.__administrator_repository.get_by_email(email)
         if administrator.status == Administrator.Status.BLOCKED:
             raise AdministratorBlockedException()
+        if is_admin and administrator is not Administrator.Role.ADMINISTRATOR:
```
* Только надо `administrator.role`
* Проверяем несколько эндпойтов, а ошибка одна "У Вас нет прав на отправку приглашений другим администраторам.". Предлагаю оставить "У Вас нет прав на это действие"?

Hrushon replied:
поспешишь - людей насмешишь

Hrushon committed:
fix(api): changed administrator invitation service

Version 4.
Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.

AntonZelinsky commented on src/core/services/authentication_service.py:
```diff
@@ -81,6 +82,8 @@ async def get_current_active_administrator(self, token: str) -> Administrator:
         if administrator.status == Administrator.Status.BLOCKED:
             raise AdministratorBlockedException()
+        if is_admin and administrator.role is not Administrator.Role.ADMINISTRATOR:
+            raise AdministratorInviteError()
```
Кажется, тут нужно форбиден исключение, так как иенно по этой причине отказ, такой админ есть, но нет прав.

Hrushon replied:
`AdministratorInviteError` как раз с форбиденом

AntonZelinsky commented on src/api/routers/administrator_invitation.py:
```diff
@@ -47,7 +47,7 @@ async def create_and_send_invitation(
         invitation_data: AdministratorInvitationRequest,
         token: HTTPAuthorizationCredentials = Depends(HTTPBearer()),
     ) -> Any:
-        await self.authentication_service.get_current_active_administrator(token.credentials)
+        await self.authentication_service.get_current_active_administrator(token.credentials, is_admin=True)
```
Получать токен тут смысла вообще нет. Создай две функции, одна будет авторизовывает админов, другая всех остальных и через депенд инжектай её. Только один раз на уровне класса, а не в каждом случае, если везде используется одна и та же проверка.

Hrushon replied:
> Получать токен тут смысла вообще нет. Создай две функции, одна будет авторизовывает админов, другая всех остальных и через депенд инжектай её. Только один раз на уровне класса, а не в каждом случае, если везде используется одна и та же проверка.

@AntonZelinsky Такое чувство, что я ничего не понял, но в итоге мне необходимо:
1. создать две функции выше - обе должны получать токен из хедера реквеста. Одна из этих функций проверяет `админ` это или `эксперт` и во втором случае райзит ошибку с форбиденом.
2. создать на уровне объекта класса свойство, в это свойство через депендс присвоить реализованную выше функцию с проверкой `админ/эксперт`

По итогу у меня несколько вопросов - помогите разобраться:
1. один из эндпоинтов `get_invitation_by_token` не требует аутентификации, как с ним быть не знаю
2. про фразу `Получать токен тут смысла вообще нет.`  - правильно ли понял, что в функции аутентификации `админ/эксперт` мне все равно придется реализовать получение токена, потому что без него я больше не знаю как идентифицировать пользователя
3. зачем вторая функция для аутентификации не админов

Hrushon replied:
Реализовал через депендс. Но, конечно, если бы не один эндпоинт, который не требует аутентификации. Все работает, кроме него - просит авторизоваться, хотя этого просить не должен.
Не знаю как обойти это, чтобы при создании класса проверять на какой эндпоинт приходит запрос и, соответственно, создавать либо не создавать свойство с авторизацией в CBV.

Kirill-Drozdov commented on src/core/services/authentication_service.py:
```diff
@@ -81,6 +82,8 @@ async def get_current_active_administrator(self, token: str) -> Administrator:
         administrator = await self.__administrator_repository.get_by_email(email)
         if administrator.status == Administrator.Status.BLOCKED:
             raise AdministratorBlockedException()
+        if is_admin and administrator.role is not Administrator.Role.ADMINISTRATOR:
```
Корутина get_current_active_administrator используется не только при работе с приглашениями. Проверь у себя локально, смогут ли эксперты работать с отчетами, сменами и выполнять прочие действия согласно бизнес-логике?

Hrushon replied:
Поэтому флаг is admin со значением по умолчанию и добавили

Kirill-Drozdov replied:
Понял. Я просто сейчас похожим багом занимаюсь. Тут так получилось, что мы почти одно и то же делаем, а задачи в Notion 2 разные. Вот моя: https://www.notion.so/127-0e9ac4f376254044abe278d2d5f1ace4?pvs=4
Я только на ревью хотел отправлять и заметил, что у нас много пересечений в коде. Возможны конфликты при мёрже.

Hrushon replied:
По твоему багу уже реализовано решение в моем ПР по замечаниям ревью.
Так что можешь ставить статус ревью и брать след задачу

Kirill-Drozdov replied:
Решение реализовано, но за рамками задачи в Notion. Там было только про активацию деактивированного приглашения. Я взял задачу с правами на деактивацию приглашения и тоже потратил на неё своё время. Да, было бы вообще логичней обе эти задачи объединить в одну в Notion, но мы работаем с тем, что есть. Я уточню по этому вопросу у старших в команде, в общем.

Hrushon committed:
fix(api): changed administrator invitation service

Version 5.
Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.
Merge branch 'develop' of github.com:Studio-Yandex-Practicum/lomaya_baryery_backend into fix/invitations_reactivate

Hrushon committed:
fix(api): changed administrator invitation service

Version 5.
Minor changes have been made.
A new error has been added.
And a new function has been added to the `AuthenticationService` class.
Fix conflicts.

ADesBiysk reviewed (commented):
Антон предлагает не использовать функцию get_current_active_administrator для проверки, что пользователь авторизован и для проверки роли. Это плохо, поскольку исходное назначение этой функции получение сущности администратора и проверка, что он не заблокирован. 
Нужно добавить 2 отдельные функции, которые или ничего не возвращают или вызывают исключение (и в БД надо не получать запись, а использовать exist):
1. Проверяет по информации в токене, что в базе существует незаблокированнный администратор с ролью администратор (не очень звучит, но не знаю как по другому сказать). Если его нет, то вызывает исключение. Она будет использоваться когда эндпойнтом может пользоваться только админ, но не эксперт.
2. Проверяет, что есть незаблокированный администратор с любой ролью. Использоваться будет когда эндпойнтом может пользоваться и администратор и эксперт.
По поводу использования в зависимостях. Это можно делать когда для каждого эндпойнта нужны одинаковые роли. Если нет, то прописываем в каждом нужные.

Hrushon committed:
fix(api): changed administrator invitation service

Version 6.
Minor changes have been made.
A new error has been added.
And a two new functions has been added to the `AuthenticationService` class and `AdministratorRepository` class.

ADesBiysk commented on src/core/services/authentication_service.py:
```diff
@@ -87,3 +87,29 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
+
+    async def check_is_active_administrator(self, token: str) -> None:
+        """Аутентифицировать активного администратора."""
+        try:
```
В 3 функциях проверка токена полностью повторяется. Нужно её вынести в отдельный приватный метод.

ADesBiysk commented on src/core/services/authentication_service.py:
```diff
@@ -87,3 +87,29 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
+        if not check_result:
+            raise exceptions.AdministratorBlockedError
+
+    async def check_is_active_superadministrator(self, token: str) -> None:
```
Название неудачное. Суперадминистратор у нас уже есть, и здесь проверка не на него. 
Плохо, что у нас есть администраторы с разными ролями и может быть администратор с ролью администратор. 
Как вариант названия функции: check_is_authorized_profile_has_role_administrator,
check_is_authorized_profile_has_any_role.

Hrushon committed:
fix(api): changed administrator invitation service

Version 7.
Minor changes have been made.
A new error has been added.
And a two new functions has been added to the `AuthenticationService` class and `AdministratorRepository` class.

AntonZelinsky commented on src/core/services/authentication_service.py:
```diff
@@ -87,3 +96,17 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
             administrator=administrator,
         )
+
+    async def check_is_authorized_profile_has_any_role(self, token: str) -> None:
```
Не используется нигде

Hrushon replied:
Цель же была создать две функции, одна из которых авторизует всех активных администраторов, а вторая только администраторов с ролью АДМИНИСТРАТОР. Я понял, что это заготовка на будущее - для других эндпоинтов. Точнее не для эндпоинтов, в которых я фиксил баги в рамках этого ПР.

Hrushon commented on src/core/services/authentication_service.py:
```diff
@@ -87,3 +96,17 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
         )
+
+    async def check_is_authorized_profile_has_any_role(self, token: str) -> None:
+        """Аутентифицировать активного администратора по токену."""
```
```suggestion
        """Авторизовать активного администратора по токену."""
```

Hrushon commented on src/core/services/authentication_service.py:
```diff
@@ -87,3 +96,17 @@ async def refresh(self, token: str) -> AdministratorAndTokensDTO:
+            raise exceptions.AdministratorBlockedError
+
+    async def check_is_authorized_profile_has_role_administrator(self, token: str) -> None:
+        """Аутентифицировать активного администратора с ролью `АДМИНИСТРАТОР` по токену."""
```
```suggestion
        """Авторизовать активного администратора с ролью `АДМИНИСТРАТОР` по токену."""
```

Hrushon committed:
Apply suggestions from code review
